src/swap/switchain.js

```diff
@@ -0,0 +1,371 @@
+
+      // get wallet addresses for exchange
+      const [fromAddress, toAddress] = await Promise.all([
+        getAddress(fromWallet, fromCurrencyCode),
+        getAddress(toWallet, toCurrencyCode)
+      ])
+
+      // Check for supported currencies, even if we aren't activated:
+      const json: SwitchainOfferResponse = await swHttpCall(
+        '/offer',
+        'GET',
+        null,
+        {
+          pair,
+          orderIdSeed: toAddress
+        }
+      )
+      const {
+        maxLimit,
+        minLimit,
+        minerFee,
+        quote,
+        signature,
+        expiryTs,
+        orderId
+      } = json
+
+      // get native min max limits
+      const [nativeMax, nativeMin] = await Promise.all([
+        fromWallet.denominationToNative(maxLimit.toString(), fromCurrencyCode),
+        fromWallet.denominationToNative(minLimit.toString(), fromCurrencyCode)
+      ])
+
+      // determine amount param for order creation
+      const [fromAmount, toAmount] = await Promise.all([
+        fromWallet.nativeToDenomination(nativeAmount, fromCurrencyCode),
+        toWallet.nativeToDenomination(nativeAmount, toCurrencyCode)
+      ])
+
+      quoteForFrom = quoteFor === 'from'
+      // check for min / max limits
+      if (quoteForFrom) {
+        if (lt(nativeAmount, nativeMin)) {
+          throw new SwapBelowLimitError(swapInfo, nativeMin)
+        }
+        if (gt(nativeAmount, nativeMax)) {
+          throw new SwapAboveLimitError(swapInfo, nativeMax)
+        }
+      } else {
+        const toAmountInFrom = div(toAmount, quote, 8)
+
+        if (lt(toAmountInFrom, minLimit)) {
+          throw new SwapBelowLimitError(swapInfo, nativeMin)
+        }
+
+        if (gt(toAmountInFrom, maxLimit)) {
+          throw new SwapAboveLimitError(swapInfo, nativeMax)
+        }
+      }
+
+      // order creation body
+      const quoteAmount = quoteForFrom ? { fromAmount } : { toAmount }
+      orderCreationBody = {
+        pair,
+        toAddress,
+        refundAddress: fromAddress,
+        signature,
+        orderId,
+        ...quoteAmount
+      }
+
+      // plugin output creation
+      fromNativeAmount = nativeAmount
+      toNativeAmount = await toWallet.denominationToNative(
+        sub(mul(quote, fromAmount), minerFee),
+        toCurrencyCode
+      )
+      if (!quoteForFrom) {
+        fromNativeAmount = await fromWallet.denominationToNative(
+          div(add(toAmount, minerFee), quote, 8),
+          fromCurrencyCode
+        )
+        toNativeAmount = nativeAmount
+      }
+
+      const destinationAddress = toAddress
+      const expirationDate = new Date(expiryTs * 1000)
+      const isEstimate = false
+
+      // create preliminary tx using our recieve address to calculate a networkFee
+      const spendInfo: EdgeSpendInfo = {
+        currencyCode: fromCurrencyCode,
+        spendTargets: [
+          { nativeAmount: fromNativeAmount, publicAddress: fromAddress }
```

Comment:
@peachbits Please test Ethereum with this. We might get "cannot send to your own address" errors on this line.

Reply:
This is the only remaining issue. As this is only a temporary `EdgeSpendInfo`, is there a way to derive a new address to be used here?

Otherwise I can think of hardcoding an address for ETH and ERC20. Any other ideas>

Reply:
Using a hard-coded addresses is the only solution we've been able to come up with when we encounter this problem ourselves. Since Ethereum is balance-based, each wallet only has a single address, so there is no way to derive another one like we can in Bitcoin.